feat(node): Add `@sentry/node/preload` hook

mydea opened the pull request:
This PR adds a new way to initialize `@sentry/node`, which allows to use the SDK with performance instrumentation even if you cannot (for whatever reason) call `Sentry.init()` at the very start of your app.

## CJS usage

In CommonJS mode, you can run the SDK like this:

```bash
node --require @sentry/node/preload ./app.js
```

```js
// app.js
const express = require('express');
const Sentry = require('@sentry/node');

const dsn = await getSentryDsn();
Sentry.init({ dsn });

// express is instrumented even though we initialized Sentry late
```

## ESM usage

in ESM mode, you can run the SDK like this:

```bash
node --import @sentry/node/preload ./app.mjs
```

```js
// app.mjs
import express from 'express';
import * as Sentry from '@sentry/node';

const dsn = await getSentryDsn();
Sentry.init({ dsn });

// express is instrumented even though we initialized Sentry late
```

## Configuration options

This script will by default preload all opentelemetry instrumentation. You can choose to instrument only specific packages like this:

```bash
SENTRY_PRELOAD_INTEGRATIONS="Http,Express,Graphql" --import @sentry/node/preload ./app.mjs
```

You can also enable debug logging for the script via `SENTRY_DEBUG=true`.

## Manually preloading

It is also possible to manually call `preloadOpenTelemetry()` to achieve the same thing. For example, in a CJS app you could do the following thing if you want to initialize late but don't want to use `--require`:

```js
// preload.js
const Sentry = require('@sentry/node');
Sentry.preloadOpenTelemetry();

// app.js
// call this first, before any other requires!
require('./preload.js');
// Then, other stuff
const express = require('express');
const Sentry = require('@sentry/node');

const dsn = await getSentryDsn();
Sentry.init({ dsn });
```

AbhiPrasad commented:
What happens if a user manually initializes OpenTelemetry? Can they still delay calling `Sentry.init`?

timfish commented:
How does this differ from using `--import @sentry/node/import`? I can see the code differenced but why is this required?

I assumed `--import @sentry/node/import` would load `import-in-the-middle` and then import order in the app doesn't matter, it still instruments libraries.

mydea commented:
> How does this differ from using `--import @sentry/node/import`? I can see the code differenced but why are these differences required?
> 
> I assumed `--import @sentry/node/import` would load `import-in-the-middle` and then import order in the app doesn't matter, it still instruments libraries.

In our tests, import/execution order sadly still mattered. If something was imported before we added the OTEL instrumentation, it did not work :( Actually, the test is not "good" because in this specific test stuff does work. I updated this now, it fails when something is used before Sentry is run. E.g.:

```js
import express from 'express';

const app = express();
// setup app...

// somewhere later
setTimeout(function() {
  Sentry.init(...);
}, 1000);
```

In this case, express is already uninstrumented before we run init :/

> What happens if a user manually initializes OpenTelemetry? Can they still delay calling Sentry.init?

This should all still work, basically this does nothing except setup the instrumentations, it does not actually setup otel itself (e.g. no span processor etc. is created yet).

andreiborza commented:
> How does this differ from using `--import @sentry/node/import`? I can see the code differenced but why are these differences required?
> 
> I assumed `--import @sentry/node/import` would load `import-in-the-middle` and then import order in the app doesn't matter, it still instruments libraries.

If I understood correctly, some people didn't have a DSN ready by the time we need to import/init everything for instrumentation to work. This would be a workaround.

mydea commented:
> > How does this differ from using `--import @sentry/node/import`? I can see the code differenced but why are these differences required?
> > I assumed `--import @sentry/node/import` would load `import-in-the-middle` and then import order in the app doesn't matter, it still instruments libraries.
> 
> If I understood correctly, some people didn't have a DSN ready by the time we need to import/init everything for instrumentation to work. This would be a workaround.

You _can_ use `@sentry/node/import` without DSN, which does setup up import-in-the-middle correctly. but this still requires the otel instrumentation to be _loaded_ before a module is used :/

timfish commented:
> ```js
> // somewhere later
> setTimeout(function() {
>   Sentry.init(...);
> }, 1000);
> ```
> In this case, express is already uninstrumented before we run init :/

Ah that does make sense for that case. 

Is there any reason not to encourage using an `instrument.js` instead?

```
node --require ./instrument.js app.js
```

```ts
// instrument.js
const Sentry = require('@sentry/node');
const dsn = await getSentryDsn();
Sentry.init({ dsn });

// app.js
const express = require('express');
```

mydea commented:
> > ```js
> > // somewhere later
> > setTimeout(function() {
> >   Sentry.init(...);
> > }, 1000);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this case, express is already uninstrumented before we run init :/
> 
> Ah that does make sense for that case.
> 
> Is there any reason not to encourage using an `instrument.js` instead?
> 
> ```
> node --require ./instrument.js app.js
> ```
> 
> ```ts
> // instrument.js
> const Sentry = require('@sentry/node');
> const dsn = await getSentryDsn();
> Sentry.init({ dsn });
> 
> // app.js
> const express = require('express');
> ```

Yeah, we generally do recommend that, but this is not possible if users cannot init at this point - e.g. if they load their DSN from somewhere else (something that we don't recommend, but there _are_ people out there that have setups like this...) 

So this is really just an alternative way to get stuff running for these people, and will def. _not_ be the recommended way to init sentry, but just an escape hatch!

timfish reviewed (approved):
Ah yes understood!

My only minor concern is that we're racking up quite a few different ways to initialise the SDK.

AbhiPrasad reviewed (approved):
> My only minor concern is that we're racking up quite a few different ways to initialise the SDK.

yeah I also feel this, but I think this is an important thing to expose, so let's release. Maybe we can iterate and combine `@sentry/node/preload` and `@sentry/node/import` somehow 🤔.

mydea commented:
> > My only minor concern is that we're racking up quite a few different ways to initialise the SDK.
> 
> yeah I also feel this, but I think this is an important thing to expose, so let's release. Maybe we can iterate and combine `@sentry/node/preload` and `@sentry/node/import` somehow 🤔.

We could, eventually, make import be a variant of preload with some env var set 🤔 but let's get this out for now and tweak stuff later as needed!

jeengbe commented:
What's a use case for this? I can't imagine any situation in which you don't have control over the entry point of your code. Maybe with frameworks like Next?
